### PR TITLE
Add digits VAE tutorial

### PR DESCRIPTION
**Pull request by jakevdp**
This is a followup to the getting started tutorial.

Rendered preview of the tutorial here: https://jax-ai-stack--59.org.readthedocs.build/en/59/digits_vae.html

### CODE REVIEW

File: docs/digits_vae.md [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+    format_version: 0.13
+    jupytext_version: 1.15.2
+kernelspec:
+  display_name: Python 3
+  name: python3
+---
+
++++ {"id": "47OmRSTR1dJU"}
+
+# Diving deeper: a Variational autoencoder (VAE) model
```
Comment (skye) [inline docs/digits_vae.md:16]:
```suggestion
# jax.jit and debugging: variational autoencoder (VAE) model
```

Or something that mentions this will cover jit + debugging?

File: docs/digits_vae.md [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+
+with jax.debug_nans(True):
+  for epoch in range(2001):
+    train_step(model, optimizer, images_train)
+```
```
Comment (skye) [inline docs/digits_vae.md:239]: `Invalid nan value encountered in the output of a C++-jit/pmap function. Calling the de-optimized version.`

Not for this PR, I feel like we can improve this message (e.g. why does it mention C++? And we have shard_map now)

Comment (skye) [reply docs/digits_vae.md:239]: Also I wonder if we can somehow make the stack trace shorter... I'm guessing probably not, but why are we reraising the exception so many times?

Also `invalid value (nan) encountered in pjit` needs to be updated

File: docs/digits_vae.md [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+train_step(model, optimizer, images_train)
+```
+
++++ {"id": "mawnZtF8wxu9"}
+
+But here rather than printing the value, we're getting some kind of `Traced` object. You'll encounter this frequently when inspecting the progress of JAX programs: tracers are the mechanism that JAX uses to implement transformations like `jit` and `grad`, and you can read more about them in [JAX Key Concepts: Tracing](https://jax.readthedocs.io/en/latest/key-concepts.html#tracing).
+
+For our purposes, the workaround is to use another tool from the [JAX debugging](https://jax.readthedocs.io/en/latest/debugging/index.html#interactive-inspection-with-jax-debug) link above: namely `jax.debug.print`, which lets us print runtime values even when they're traced:
```
Comment (skye) [inline docs/digits_vae.md:317]: I don't think JAX debugging is linked above?

**Review by skye — COMMENTED** [review]
I didn't do a super careful read but overall this looks AMAZING

File: docs/digits_vae.md (hunk 3 above)
Comment (jakevdp) [reply docs/digits_vae.md:318]: It's linked on line 221, but with a different link text. I could make them match – WDYT?

File: docs/digits_vae.md (hunk 2 above)
Comment (jakevdp) [reply docs/digits_vae.md:240]: Yeah agreed this error is really hard to parse. Maybe some of it could be fixed by using `tracebackutil.register_exclusion` within parts of NNX

**Comment by 8bitmp3** [comment]
Thank you @jakevdp . Will review in the next round after you merge the PR with @skye .

File: docs/digits_vae.md (hunk 3 above)
Comment (skye) [reply docs/digits_vae.md:318]: Ah yeah making them match sounds good. I tried searching for "JAX debugging" to find where it was mentioned above because I didn't remember seeing it.

**Review by skye — APPROVED** [review]
LGTM btw. I didn't approve before because I assumed you were just merge it when you felt it was ready :)

File: docs/digits_vae.md (hunk 3 above)
Comment (jakevdp) [reply docs/digits_vae.md]: text updated

File: docs/digits_vae.md (hunk 2 above)
Comment (jakevdp) [reply docs/digits_vae.md:240]: cc/ @cgarciae – can we take a look at how tracebacks interact with NNX transformations?

**Comment by rcrowe-google** [comment]
Is the stack trace under "Debugging NaNs in JAX" supposed to be that long?

**Comment by jakevdp** [comment]
Yes, that's the current stack trace that JAX generates. See Skye's comments above – there are potentially ways we could update JAX and/or flax to address this.

File: docs/digits_vae.md [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+images_train, images_test = map(jnp.asarray, splits)
+
+print(f"{images_train.shape=}")
+print(f"{images_test.shape=}")
+```
+
++++ {"id": "2_Q16JRyrW7V"}
+
+The dataset comprises 1800 images, each represented by an 8x8 pixel grid.
+To see a visualization of this data, refer to [loading the data](https://jax-ai-stack.readthedocs.io/en/latest/getting_started_with_jax_for_AI.html#loading-the-data) in the previous tutorial.
```
Comment (melissawm) [inline docs/digits_vae.md:49]: Unfortunately I can't add suggestions here, but you could use 
```
To see a visualization of this data, refer to [loading the data](getting_started_with_jax_for_AI.ipynb#loading-the-data) in the previous tutorial.
```
to create a link to the correct version of the target page (using hard links may cause the readthedocs page to be redirected to latest instead of stable and create unnecessary warnings).

File: docs/digits_vae.md [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+    x = nnx.selu(self.layer1(x))
+    x = nnx.selu(self.layer2(x))
+    return self.layer3(x)
+```
+
++++ {"id": "3DwMBvoksOmG"}
+
+In this network we had one output per class, and the loss function was designed such that once trained, the output corresponding to the correct class would return the strongest signal, thus predicting the correct label in upwards of 95% of cases.
+
+In this VAE example we use similar building blocks to instead output a small probabilisitic model representing the data. While classic `VAE` is generally based on convolutional layers, we use linear layers for simplicity. The sub-network that produces this probabilistic encoding is our `Encoder`:
```
Comment (melissawm) [inline docs/digits_vae.md:81]: Suggested change:
```
In this VAE example we use similar building blocks to instead output a small probabilistic model representing the data. While classic `VAE` is generally based on convolutional layers, we use linear layers for simplicity. The sub-network that produces this probabilistic encoding is our `Encoder`:
```

File: docs/digits_vae.md [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+    z = self.linear1(z)
+    z = jax.nn.relu(z)
+    logits = self.linear2(z)
+    return logits
+```
+
++++ {"id": "0QaT-KY6npSc"}
+
+Now the full VAE model is a single network built from the encoder and decoder.
+It returns both the reconstructed image and then internal latent space model:
```
Comment (melissawm) [inline docs/digits_vae.md:131]: Suggested change:
```
It returns both the reconstructed image and the internal latent space model:
```

File: docs/digits_vae.md [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+```
+
++++ {"id": "xIm9Yi5YoIxN"}
+
+Next is the loss function – there are two components to the model that we want to ensure:
+
+1. the `logits` output faithfully reconstruct the input image.
+2. the model represented by `mean` and `std` faithfully represents the "true" latent distribution.
+
+VAE uses a loss function based on the [Evidence lower bound](https://en.wikipedia.org/wiki/Evidence_lower_bound) to quantify theset two goals in a single loss value:
```
Comment (melissawm) [inline docs/digits_vae.md:166]: Suggested change:
```
VAE uses a loss function based on the [Evidence lower bound](https://en.wikipedia.org/wiki/Evidence_lower_bound) to quantify these two goals in a single loss value:
```

File: docs/digits_vae.md [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+model = VAE(
+  image_shape=(8, 8),
+  hidden_size=32,
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+
+for epoch in range(301):
```
Comment (melissawm) [inline docs/digits_vae.md:275]: Should this be `501`?

File: docs/digits_vae.md [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+  loss = train_step(model, optimizer, images_train)
+  if epoch % 50 == 0:
+    print(f'Epoch {epoch} loss: {loss}')
+```
+
++++ {"id": "Jk_wdvqpurTG"}
+
+It looks like our loss value is decreasing toward negative infinity until the point where the values are no longer well-represented by floating point math.
+
+At this point, we may wish to expect the values within the loss function itself to see where the diverging loss might be coming from.
```
Comment (melissawm) [inline docs/digits_vae.md:285]: I think you mean
```
At this point, we may wish to inspect the values within the loss function itself to see where the diverging loss might be coming from.
```

**Review by melissawm — COMMENTED** [review]
This is a very nice read, I only saw a couple of typos and have a suggestion for a link. Cheers!

File: docs/digits_vae.md (hunk 4 above)
Comment (jakevdp) [reply docs/digits_vae.md:49]: I prefer the full link so that it works when viewing the notebook directly (if we were just concerned about links in the rendered HTML, I'd use sphinx cross-references instead).

File: docs/digits_vae.md (hunk 5 above)
Comment (jakevdp) [reply docs/digits_vae.md:81]: Thanks! Done in #75

File: docs/digits_vae.md (hunk 9 above)
Comment (jakevdp) [reply docs/digits_vae.md:285]: Done in #75

File: docs/digits_vae.md (hunk 8 above)
Comment (jakevdp) [reply docs/digits_vae.md:275]: Done in #75

File: docs/digits_vae.md (hunk 6 above)
Comment (jakevdp) [reply docs/digits_vae.md:131]: Done in #75
